test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+
+            done();
+          });
+        });
+    });
+  });
+
+  describe('User.logout', function() {
+    it('Logout a user by providing the current accessToken id (using node)', function(done) {
+      login(logout);
+
+      function login(fn) {
+        User.login({email: 'foo@bar.com', password: 'bar'}, fn);
+      }
+
+      function logout(err, accessToken) {
+        User.logout(accessToken.id, verify(accessToken.id, done));
+      }
+    });
+
+    it('Logout a user by providing the current accessToken id (using node) - promise variant',
+    function(done) {
+      login(logout);
+
+      function login(fn) {
+        User.login({email: 'foo@bar.com', password: 'bar'}, fn);
+      }
+
+      function logout(err, accessToken) {
+        User.logout(accessToken.id)
+          .then(function() {
+            verify(accessToken.id, done);
+          })
+          .catch(done(err));
+      }
+    });
+
+    function verify(token, done) {
+      assert(token);
+
+      return function(err) {
+        if (err) return done(err);
+
+        AccessToken.findById(token, function(err, accessToken) {
+          assert(!accessToken, 'accessToken should not exist after logging out');
+
+          done(err);
+        });
+      };
+    }
+
+    it('Logout a user from user model 1 without logout user from model 2', function(done) {
+      User.login({email: 'foo@bar.com', password: 'bar'}, function(err, accessToken1) {
+        User.login({email: 'foo@bar.com', password: 'bar'}, function(err, accessToken2) {
+          User.logout(accessToken1.id, function(err, res) {
+            AccessToken.findById(accessToken1.id, function(err, accessToken) {
+              assert(!accessToken, 'accessToken should not exist after logging out user1');
+              AccessToken.findById(accessToken2.id, function(err, accessToken2) {
+                assert(accessToken2, 'accessToken2 should exist after logging out user1');
+                done();
+              });
+            });
+          });
+        });
+      });
+    });
+  });
+
+  describe('Password Reset', function() {
+    describe('User.resetPassword(options, cb)', function() {
+      var options = {
+        email: 'foo@bar.com',
+        redirect: 'http://foobar.com/reset-password',
+      };
+
+      it('Creates a temp accessToken to allow a user to change password', function(done) {
+        var calledBack = false;
+
+        User.resetPassword({
+          email: options.email,
+        }, function() {
+          calledBack = true;
+        });
+
+        User.once('resetPasswordRequest', function(info) {
+          assert(info.email);
+          assert(info.accessToken);
+          assert(info.accessToken.id);
+          assert.equal(info.accessToken.ttl / 60, 15);
+          assert(calledBack);
+          info.accessToken.user(function(err, user) {
+            if (err) return done(err);
+            assert.equal(user.email, options.email);
+
```

Comment:
In the context of polymorphic relation, I think it's important to verify that the access token is created for the correct user model, and that `info.accessToken.user` return the correct user model too. Thoughts?